File: framework/test2/validate/list/List.scala
import soda.scala.unittest._

import scala.reflect.runtime.universe.typeOf

import soda.scala.leetcode._

import collection.mutable
import collection.mutable.{ArrayBuffer, ArrayDeque, PriorityQueue}
import util.control.Breaks.{break, breakable}
object Solution {
  def permutation(chars: Array[Char], n: Int): Array[String] = {
    val res = ArrayBuffer[String]()
    val buf = Array.fill[Char](n)(' ')
    solve(chars, 0, buf, 0, res)
    res.toArray
  }

  private def solve(chars: Array[Char], i: Int, buf: Array[Char], j: Int, res: ArrayBuffer[String]): Unit = {
    if (j == buf.length) {
      res += new String(buf)
      return
    }
    for (k <- i until chars.length) {
      var temp = chars(i)
      chars(i) = chars(k)
      chars(k) = temp
      buf(j) = chars(i)
      solve(chars, i+1, buf, j+1, res)
      temp = chars(i)
      chars(i) = chars(k)
      chars(k) = temp
    }
  }
}

class List {
  def get(): () => Unit = {
    () => {
      val work = GenericTestWork.create2(Solution.permutation)
      // val work = GenericTestWork.forStruct(typeOf[STRUCT])
      // val work = TestWork.forObject(typeOf[Solution], "METHOD")
      // val work = TestWork.forInstance(typeOf[Solution], "METHOD")
      // val work = TestWork.createN((...)=>R)
      // val work = TestWork.forStruct(typeOf[STRUCT])
      work.setValidator(Validators.forArray[String](false))
      work.compareSerial = true
      work.run()
    }
  }
}

object List extends App {
  new List().get().apply()
}

File: framework/test2/validate/list/List.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.Validators;
import soda.unittest.work.TestWork;


class Solution {
    public String[] permutation(char[] chars, int n) {
        var res = new ArrayList<String>();
        char[] buf = new char[n];
        solve(chars, 0, buf, 0, res);
        return res.toArray(new String[0]);
    }

    private void solve(char[] chars, int i, char[] buf, int j, java.util.List<String> res) {
        if (j == buf.length) {
            res.add(new String(buf));
            return;
        }
        for (int k = i; k < chars.length; ++k) {
            char temp = chars[i];
            chars[i] = chars[k];
            chars[k] = temp;
            buf[j] = chars[i];
            solve(chars, i+1, buf, j+1, res);
            temp = chars[i];
            chars[i] = chars[k];
            chars[k] = temp;
        }
    }
}

public class List implements Supplier<TestWork> {

    @Override
    public TestWork get() {
        var work = new TestWork(new Solution(), "permutation");
        // var work = TestWork.forStruct(Struct.class);
        work.setValidator(Validators.forArray(String.class, false));
        work.setCompareSerial(true);
        return work;
    }

    public static void main(String[] args) throws Exception {
        new List().get().run();
    }

}
